### ISSUE TITLE
SchemaSpy for Informix: primary key columns list and foreign key columns lists not sorted correctly

### ISSUE THREAD

**Issue by begooden-it** [schemaspy/schemaspy#638]
In the .xml file, on informix 14.10, and  on previous versions, SchemaSpy does not always print the primary keys and the foreign keys columns list in the same order as their respective indexes: 
Since I am using the schemaspy xml file as a datasource for a project of mine, (code generator), it often happens that listing the primary key columns for the parent tables does not render the same order for the foreign order columns.

## Expected Behavior

the order of the list of columns of the primary key should be strictly and always the same as the order of the matching foreign key's columns. The reference should be the order of the index used by the primary key and the foreign key

## Steps to Reproduce (for bugs)

1. will try to push an xml file to you where it is obvious
2.
3.
4.

## Context

* Database type (e.g. MySQL):
* Database driver version (is you use a special one):
* Your DDL (sample how to reproduce the issue):
* Full command including arguments how you run SchemaSpy:

## Your Environment

* Version used:
* Environment name and version (e.g. Chrome 39):
* Operating System and version (e.g. Windows 10):
* Java version:

**Comment by npetzall** [comment]
Duplicate of #380, might be possible to fix with override query that has an order by.

**Comment by begooden-it** [comment]
Hi Nils,





That would definitely help 😊







*De :* Nils Petzäll <notifications@github.com>
*Envoyé :* lundi 27 janvier 2020 22:12
*À :* schemaspy/schemaspy <schemaspy@noreply.github.com>
*Cc :* begooden-it <eric.vercelletto@begooden-it.com>; Author <
author@noreply.github.com>
*Objet :* Re: [schemaspy/schemaspy] SchemaSpy for Informix: primary key
columns list and foreign key columns lists not sorted correctly (#638)



Duplicate of #380 <https://github.com/schemaspy/schemaspy/issues/380>,
might be possible to fix with override query that has an order by.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/schemaspy/schemaspy/issues/638?email_source=notifications&email_token=AEO74TBVO2FL3DZMRHQIM3TQ75E2BA5CNFSM4KMIC7U2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKBCKHY#issuecomment-578954527>,
or unsubscribe

**Comment by begooden-it** [comment]
Can I do that in my Informix. Properties file ?



If yes, what should I do?



Thanks



Eric



*De :* Nils Petzäll <notifications@github.com>
*Envoyé :* lundi 27 janvier 2020 22:12
*À :* schemaspy/schemaspy <schemaspy@noreply.github.com>
*Cc :* begooden-it <eric.vercelletto@begooden-it.com>; Author <
author@noreply.github.com>
*Objet :* Re: [schemaspy/schemaspy] SchemaSpy for Informix: primary key
columns list and foreign key columns lists not sorted correctly (#638)



Duplicate of #380 <https://github.com/schemaspy/schemaspy/issues/380>,
might be possible to fix with override query that has an order by.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/schemaspy/schemaspy/issues/638?email_source=notifications&email_token=AEO74TBVO2FL3DZMRHQIM3TQ75E2BA5CNFSM4KMIC7U2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKBCKHY#issuecomment-578954527>,
or unsubscribe

**Comment by npetzall** [comment]
I haven't had the time to really look into this, but primary Keys can have an override sql. 

If I remember correctly the jdbc api says that it sorts by name instead of seqno. 
So with a custom query that sorts by seqno or similar should solve the problem. 

When I created the override I added the same columns as returned by jdbc. 

selectPrimaryKeysSql=
Fetch table PKs, expected columns:
TABLE_CAT, TABLE_SCHEM, TABLE_NAME, COLUMN_NAME, KEY_SEQ, PK_NAME

**Comment by npetzall** [comment]
@begooden-it I've fixed the issue #380 which relates to PK however I'm a bit unsure about the FK issue.

**Comment by npetzall** [comment]
@begooden-it after #380 have you had any issue or any reproducible case.

I've quickly read the documentation and from the functions we use in JDBC it seems that PK is order by name but other are by sequence number with in the composite key.

So the difference you've seen might only have been the incorrect PK. But if there is an issue it would save me a lot of time if you could provide an example.

**Comment by begooden-it** [comment]
Hi Nils,







Sorry for the delay, messy times here …



I’ll take a look ASAP and provide a testcase



Thanks



Eric



*De :* Nils Petzäll <notifications@github.com>
*Envoyé :* samedi 14 mars 2020 20:51
*À :* schemaspy/schemaspy <schemaspy@noreply.github.com>
*Cc :* begooden-it <eric.vercelletto@begooden-it.com>; Mention <
mention@noreply.github.com>
*Objet :* Re: [schemaspy/schemaspy] SchemaSpy for Informix: primary key
columns list and foreign key columns lists not sorted correctly (#638)



@begooden-it <https://github.com/begooden-it> after #380
<https://github.com/schemaspy/schemaspy/issues/380> have you had any issue
or any reproducible case.

I've quickly read the documentation and from the functions we use in JDBC
it seems that PK is order by name but other are by sequence number with in
the composite key.

So the difference you've seen might only have been the incorrect PK. But if
there is an issue it would save me a lot of time if you could provide an
example.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/schemaspy/schemaspy/issues/638#issuecomment-599127567>,
or unsubscribe

**Comment by npetzall** [comment]
@begooden-it no worries, the situation is pretty crazy in Europe right now. Seems like we in Sweden are getting there too.

**Comment by begooden-it** [comment]
Hi again,



Taking benefit of this heavy silence in French streets, here some evidence
material



Extract from the .xml file for table customer

      <table catalog="kandoodb" name="customer" numRows="105" remarks=""
schema="informix" type="TABLE">

         <primaryKey column="cmpy_code" sequenceNumberInPK="1"/>

         <primaryKey column="cust_code" sequenceNumberInPK="2"/>



And dbschema output of the same table



create unique index "informix".u_customer on "informix".customer
(cust_code,cmpy_code) using btree ;

alter table "informix".customer add constraint primary key
(cust_code,cmpy_code) constraint "informix".pk_customer  ;



--------------------------------------------------------------------------------------------------------------------------------

Same for the table ingroup

      <table catalog="kandoodb" name="ingroup" numRows="10" remarks=""
schema="informix" type="TABLE">

         <primaryKey column="cmpy_code" sequenceNumberInPK="1"/>

         <primaryKey column="ingroup_code" sequenceNumberInPK="2"/>

         <primaryKey column="type_ind" sequenceNumberInPK="3"/>



Dbschema

    desc_text nvarchar(40),

    primary key (cmpy_code,type_ind,ingroup_code)

  );



The order specified in sequenceNumberInPK Is sometimes different from the
order you will find in the dbschema output, which is the real order. It
seems that SchemaSpy always sorts by name, after looking at more
occurrences.



I join a zip file containing full dbschema and full .xml files



*De :* Eric Vercelletto <eric.vercelletto@begooden-it.com>
*Envoyé :* mercredi 18 mars 2020 12:39
*À :* 'schemaspy/schemaspy' <
reply@reply.github.com>;
'schemaspy/schemaspy' <schemaspy@noreply.github.com>
*Cc :* 'Mention' <mention@noreply.github.com>
*Objet :* RE: [schemaspy/schemaspy] SchemaSpy for Informix: primary key
columns list and foreign key columns lists not sorted correctly (#638)



Hi Nils,







Sorry for the delay, messy times here …



I’ll take a look ASAP and provide a testcase



Thanks



Eric



*De :* Nils Petzäll <notifications@github.com>
*Envoyé :* samedi 14 mars 2020 20:51
*À :* schemaspy/schemaspy <schemaspy@noreply.github.com>
*Cc :* begooden-it <eric.vercelletto@begooden-it.com>; Mention <
mention@noreply.github.com>
*Objet :* Re: [schemaspy/schemaspy] SchemaSpy for Informix: primary key
columns list and foreign key columns lists not sorted correctly (#638)



@begooden-it <https://github.com/begooden-it> after #380
<https://github.com/schemaspy/schemaspy/issues/380> have you had any issue
or any reproducible case.

I've quickly read the documentation and from the functions we use in JDBC
it seems that PK is order by name but other are by sequence number with in
the composite key.

So the difference you've seen might only have been the incorrect PK. But if
there is an issue it would save me a lot of time if you could provide an
example.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/schemaspy/schemaspy/issues/638#issuecomment-599127567>,
or unsubscribe

**Comment by npetzall** [comment]
This will be interesting since I think I've fixed it. But it could be an issue with the jdbc driver. Which version are you using?

If it's and issue with the jdbc driver I'll add an override sql, if not I'll try to fix my fix.

**Comment by begooden-it** [comment]
JDBC is JDBC_450.JC2



Now last time I checked was with a not so recent schemaSpy build .



Since I am using the last snapshot, let me check again. The issue was on
another server







*De :* Nils Petzäll <notifications@github.com>
*Envoyé :* mercredi 18 mars 2020 15:03
*À :* schemaspy/schemaspy <schemaspy@noreply.github.com>
*Cc :* begooden-it <eric.vercelletto@begooden-it.com>; Mention <
mention@noreply.github.com>
*Objet :* Re: [schemaspy/schemaspy] SchemaSpy for Informix: primary key
columns list and foreign key columns lists not sorted correctly (#638)



This will be interesting since I think I've fixed it. But it could be an
issue with the jdbc driver. Which version are you using?

If it's and issue with the jdbc driver I'll add an override sql, if not
I'll try to fix my fix.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/schemaspy/schemaspy/issues/638#issuecomment-600641191>,
or unsubscribe

**Comment by begooden-it** [comment]
You are right, the problem is resolved in the snapshot!



My fault



We can close



Thanks!



*De :* Nils Petzäll <notifications@github.com>
*Envoyé :* mercredi 18 mars 2020 15:03
*À :* schemaspy/schemaspy <schemaspy@noreply.github.com>
*Cc :* begooden-it <eric.vercelletto@begooden-it.com>; Mention <
mention@noreply.github.com>
*Objet :* Re: [schemaspy/schemaspy] SchemaSpy for Informix: primary key
columns list and foreign key columns lists not sorted correctly (#638)



This will be interesting since I think I've fixed it. But it could be an
issue with the jdbc driver. Which version are you using?

If it's and issue with the jdbc driver I'll add an override sql, if not
I'll try to fix my fix.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/schemaspy/schemaspy/issues/638#issuecomment-600641191>,
or unsubscribe

**Comment by npetzall** [comment]
No worries, then this was fixed by #380
